FUSETOOLS-2809 - first cut at a read-only version of the rest page

bfitzpat opened the pull request:
Refactoring UI
Cleaning up code and disabling controls for read-only display.

Signed-off-by: Brian Fitzpatrick <bfitzpat@redhat.com>

# Pull Request Checklist

After this checklist is all checked or PR provides explanations for possible pass-through, please put the label "Ready for review" on the PR.


## General

- [x] Did you use the Jira Issue number in the commit comments?
- [x] Did you set meaningful commit comments on each commit?
- [x] Did you sign off all commits for this PR? (git commit -s -m "jira issue number - your commit comment")

## Functional

- [x] Did the CI job report a successful build?
- [x] Is the non-happy path working, too?

## Maintainability

- [x] Are all Sonar reported issues fixed or can they be ignored?
- [x] Is there no duplicated code?
- [x] Are method-/class-/variable-names meaningful?

## Tests

- [ ] Are there unit-tests?
- [x] Are there integration tests (or at least a jira to tackle these)?
- [ ] Do we need a new UI test?

## Legal

- [x] Have you used the correct file header copyright comment?
- [x] Have you used the correct year in the headers of new files?

apupier commented:
If i select an operation, the editor becomes dirty and the "uri" value of the selected operation is lost

bfitzpat commented on editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/restconfiguration/RestConfigEditor.java:
```diff
@@ -0,0 +1,721 @@
+		tabFolder.setSelectionBackground(display.getSystemColor(SWT.COLOR_GRAY));
+	}
+	
+	private void buildModel() {
```
Do you have an example of what you are looking for here?

apupier replied:
I have this kind of things in mind: https://github.com/bfitzpat/fuseide/pull/13

bfitzpat commented on editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/restconfiguration/RestConfigEditor.java:
```diff
@@ -0,0 +1,721 @@
+
+		toolkit.createLabel(client, UIMessages.RestConfigEditorBindingModeLabel);
+		bindingModeCombo = new Combo(client, SWT.DROP_DOWN);
+		bindingModeCombo.add("auto"); //$NON-NLS-1$
```
It came from the Camel docs https://camel.apache.org/rest-dsl.html "Binding to POJOs Using". Not sure how to dive into the catalog to get this level of detail?

bfitzpat commented on editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/restconfiguration/RestConfigEditor.java:
```diff
@@ -0,0 +1,721 @@
+		
+		toolkit.createLabel(client, UIMessages.RestConfigEditorComponentLabel);
+		componentCombo = new Combo(client, SWT.DROP_DOWN);
+		componentCombo.add("netty-http"); //$NON-NLS-1$
```
Camel docs https://camel.apache.org/rest-dsl.html

bfitzpat commented on editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/restconfiguration/RestConfigEditor.java:
```diff
@@ -0,0 +1,721 @@
+
+	private void createColors() {
+		colorMap = new HashMap<>();
+		colorMap.put(RestConfigConstants.REST_COLOR_LIGHT_BLUE, new Color(Display.getDefault(), 235, 242, 250));
```
Do you have an example of what this might look like? I like the idea of abstracting it further.

apupier replied:
see https://github.com/bfitzpat/fuseide/pull/14
It allows to instantiate a single time each color and don't create/dispose on each editor opening. I refactored in a second commit to use maps instead of if/then/else structure

bfitzpat commented:
@apupier I think the issue has to do with the model changed listener and the property pages. Not sure how yet to handle that, but wondering if we want to disallow property page editing entirely and just go with a wizard-based approach?

apupier commented:
> @apupier I think the issue has to do with the model changed listener and the property pages. Not sure how yet to handle that, but wondering if we want to disallow property page editing entirely and just go with a wizard-based approach?

I don't like the idea to have a wizard-based approach.
I don't know how to disable the edition in the properties page. I guess it will require to modify the properties page.

apupier commented:
While trying to see what we can do to put properties in read-only I noticed that the properties displayed are not always in sync with what is displayed.
I tried to start fixing it without success. the hand-support of the selection and the usage of a different element in the editor selection than what we want to use in properties as source lost me.
here is my trial: https://github.com/apupier/jbosstools-fuse
(please note that I also update the test name as you moved it to integration, the commit can be cherry-picked separately)

apupier commented:
I removed "ready for review" as there are several feedback which need to be addressed.
Shouldn't we remove the "draft" label too?

bfitzpat commented:
@apupier and @lhein I have attempted to resolve some of the outstanding issues. 

We removed the multiple tabs and have everything in collapsible sections on the REST page now, which simplifies things a bit. Also added some of the code Aurelien was toying with as far as triggering a model refresh when the user goes between pages. 

That said I'm still not quite sure what Aurelien is after as far as a "color registry" but I have addressed some of the refresh issues when switching between pages. 

And I noticed that the model doesn't refresh consistently though going from the Source tab to the REST tab, but if you go through the Configurations tab, it works. So I'm not sure what's up there. No difference in the code encountered along the path that I can immediately point to. 

Would appreciate another set of eyes if either of you get a chance to take a look. Thanks.

apupier commented on editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/restconfiguration/RestConfigEditor.java:
```diff
@@ -0,0 +1,676 @@
+			throws PartInitException {
+		setSite(editorSite);
+		setInput(input);
+//		getSite().setSelectionProvider(this);
```
deactivating the selection completely is then providing very little information on the rest configuration. Why not for a very first iteration as the editor is hidden by default  anyway.
Would need to update the properties view to not have the previous element selection still displayed (for instance element of the selected element in the design tab before going to Rest tab)

apupier commented:
> And I noticed that the model doesn't refresh consistently though going from the Source tab to the REST tab, but if you go through the Configurations tab, it works. So I'm not sure what's up there. No difference in the code encountered along the path that I can immediately point to.

see https://github.com/bfitzpat/fuseide/pull/15 for what seems to be source of the issue. I don't understand yet why needs to be in sync here and not for the other tabs but at least it is working.

bfitzpat commented:
Updated and squashed commits to keep this branch a little cleaner, but we seem to be in better shape

bfitzpat commented:
Next step - add a read-only properties view to have something bringing enough value to users in a second iteration/PR

But I think we're ready (if not very close) to a mergeable state for this first chunk after chatting with @apupier (and big thanks to Aurelien for helping with multiple aspects of this code!)

bfitzpat commented on editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/restconfiguration/RestConfigEditor.java:
```diff
@@ -0,0 +1,613 @@
+	@Override
+	public ISelection getSelection() {
+		// TODO return selection if not null return new StructuredSelection(selection)
+		// in next iteration
```
Leaving this for the next round of changes.

bfitzpat commented on editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/restconfiguration/RestConfigEditor.java:
```diff
@@ -0,0 +1,613 @@
+				Element restElement = (Element) event.getStructuredSelection().getFirstElement();
+				// TODO call setSelection(new StructuredSelection(new CamelBasicModelElement(null, restElement)))
+				// in next iteration
+				
```
Leaving this for the next round of changes.

bfitzpat commented on editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/restconfiguration/RestConfigEditor.java:
```diff
@@ -0,0 +1,613 @@
+			if (selection != null) {
+				// TODO call setSelection (new StructuredSelection(selection) ) in next iteration
+			}
+		}
```
Leaving this for the next round of changes.

bfitzpat commented:
Built in Fuse Jenkins at least! I think we can ignore the Sonar issues for now.

apupier commented:
> I think we can ignore the Sonar issues for now.

fixing the critical one is important.

bfitzpat commented:
testPR

apupier commented on core/plugins/org.fusesource.ide.camel.model.service.core/src/org/fusesource/ide/camel/model/service/core/model/CamelContextElement.java:
```diff
@@ -252,26 +252,34 @@ protected void parseChildren() {
+			addChildElement(cme);
+		} else if (REST_CONFIGURATION_NODE_NAME.equals(tagNameWithoutPrefix) ||
+				REST_NODE_NAME.equals(tagNameWithoutPrefix)) {
+			// these are handled differently on the REST page in the editor, so let them by
```
for future iteration: I think we should be able to create the model from here instead of building it in Editor.

apupier commented on editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/internal/l10n/messages.properties:
```diff
@@ -218,6 +218,21 @@ updatePomDependenciesProgressDialogLabel=Updating Maven project file for new con
 loadingCamelFile=Loading {0} of project {1}
 pending=Pending ...
 
+editorPreferencePageTechPreviewRESTEditorPageSetting=Enable Tech Preview REST DSL editor (Requires editor restart)
```
might be nice to specify that it will be in read-only (can be done in a separated PR)

bfitzpat replied:
https://issues.jboss.org/browse/FUSETOOLS-2872

cunningt commented:
testPR

bfitzpat commented:
testPR

apupier commented:
there was a successful build for normal PR job https://fusesource-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/jbosstools-fuse_PullRequest/904/
the new PR job has been triggered by the "testPR" so we are good to merge
